src/Hamiltonians/HubbardRealSpace.jl

```diff
@@ -53,58 +53,119 @@ end
+@inline _interaction_col(a, ::Tuple{}, ::Tuple{}, ::Tuple{}, ::ModeMap, ::Tuple{}, ::CubicGrid) = 0
+@inline function _interaction_col(a, (b, bs...), (u, us...), (w, ws...), occ_a, (occ_b, occs...), g::CubicGrid)
+    return local_interaction(a, b, u, occ_a, occ_b) + _interaction_col(a, bs, us, ws, occ_a, occs, g) +
+        nearest_neighbor_interaction(onr(a), onr(b), w, g, occ_a)
+end
+@inline _interaction_col(a, ::Tuple{}, ::Tuple{}, ::Tuple{Nothing}, ::ModeMap, ::Tuple{}, ::CubicGrid) = 0
+@inline function _interaction_col(a, (b, bs...), (u, us...), w::Tuple{Nothing}, occ_a, (occ_b, occs...), g::CubicGrid)
+    return local_interaction(a, b, u, occ_a, occ_b) + _interaction_col(a, bs, us, w, occ_a, occs, g)
+end
+@inline _interaction_col(a, ::Tuple{}, ::Tuple{Nothing}, ::Tuple{}, ::ModeMap, ::Tuple{}, ::CubicGrid) = 0
+@inline function _interaction_col(a, (b, bs...), u::Tuple{Nothing}, (w, ws...), occ_a, (occ_b, occs...), g::CubicGrid)
+    return _interaction_col(a, bs, u, ws, occ_a, occs, g) +
+        nearest_neighbor_interaction(onr(a), onr(b), w, g, occ_a)
 end
-
 """
-    _interactions(addresses, interaction_matrix, occs)
+    _interactions(addresses, onsite_int_matrix, nearest_neighbour_int_matrix, occs, geometry)
 
-Compute all pairwise interactions in a tuple of `addresses`. The `interaction_matrix` sets
-the intraction strengths and `occs` holds the occupied modes of the adresses.
+Compute all pairwise interactions in a tuple of `addresses`. The `onsite_int_matrix` and
+`nearest_neighbour_int_matrix` sets the intraction strengths of the onsite interaction and
+the nearest neighbour interaction. Moreover, `occs` holds the occupied modes of the adresses
 
 The code is equivalent to the following.
 
 ```julia
 acc = 0.0
 for (i, a) in enumerate(addresses)
-    acc += local_interaction(a, interaction_matrix[i, i])
+    acc += local_interaction(a, onsite_int_matrix[i, i]) +
+        nearest_neighbour_int_matrix(a, nearest_neighbour_int_matrix[i,i],
+        occupied_mode_map(a), geometry)
     for (j, b) in enumerate(addresses[i+1:end])
-        acc += local_interaction(a, b, interaction_matrix[i, j])
+        acc += local_interaction(a, b, onsite_int_matrix[i, j]) +
+            nearest_neighbor_interaction(a, b, nearest_neighbour_int_matrix[i,j],
+            occupied_mode_map(a), geometry)
     end
 end
 return acc
 ```
 
 It is implemented recursively to ensure type stability.
 """
-@inline _interactions(::Tuple{}, ::SMatrix{0,0}, ::Tuple{}) = 0.0
-@inline function _interactions(
-    (a, as...)::NTuple{N,AbstractFockAddress}, m::SMatrix{N,N}, (occ, occs...)
-) where {N}
+@inline _interactions(::Tuple{}, ::Union{SMatrix{0,0},Nothing}, ::Union{SMatrix{0,0},Nothing},
+    ::Tuple{}, ::CubicGrid) = 0.0
+@inline function _interactions((a, as...)::NTuple{N,AbstractFockAddress},
+    m::Union{SMatrix{N,N},Nothing}, σ::Union{SMatrix{N,N},Nothing}, (occ, occs...),
```

Comment:
```suggestion
    m::Union{SMatrix{N,N},Nothing}, σ::Union{SMatrix{N,N},Nothing}, (occ, occs...),
```
I don't know why I originally named the matrix `m` (looks like a bad idea to me). It'd be better if these were `u_matrix` and `w_matrix`, or something similar so it's clearer what they represent.